### PR TITLE
Implement node banning

### PR DESCRIPTION
**Pull request by AndrejMitrovic**
This requires:

https://github.com/bpfkorea/agora/pull/106
https://github.com/bpfkorea/agora/pull/108

Note: this contains from those two PRs. Please ignore this PR until these two above are merged first, and then I'll rebase this one.

### CODE REVIEW

**Comment by AndrejMitrovic** [comment]
Unblocked, but now contains too many commits. I should split this up.

File: source/agora/network/NetworkManager.d [hunk 1]
```diff
@@ -140,9 +140,11 @@ public class NetworkManager
         // start validating and voting on the blockchain
         this.taskman.runTask(()
         {
-            while (!this.peerLimitReached())
+            while (1)
```
Comment (AndrejMitrovic) [inline source/agora/network/NetworkManager.d:143]: Why did I change this?

Comment (AndrejMitrovic) [reply source/agora/network/NetworkManager.d:143]: Ah yes, because this process is constant. I will add a comment.

Comment (Geod24) [reply source/agora/network/NetworkManager.d:143]: Walter, is that you ?

File: source/agora/network/NetworkClient.d [hunk 2]
```diff
@@ -65,17 +68,19 @@ class NetworkClient
...
             api = the API to issue the requests with
             retry = the amout to wait between retrying failed requests
             max_retries = max number of times a failed request should be retried
 
     ***************************************************************************/
 
-    public this (TaskManager taskman, Address address, API api, Duration retry,
-        size_t max_retries)
+    public this (TaskManager taskman, BanManager banman, Address address,
+        API api, Duration retry, size_t max_retries )
```
Comment (Geod24) [inline source/agora/network/NetworkClient.d]: !!!!!!!!!!!!!!!

Comment (AndrejMitrovic) [reply source/agora/network/NetworkClient.d]: OMG

**Comment by AndrejMitrovic** [comment]
Unfortunately it seems the coverage went down because of multithreading..

**Comment by AndrejMitrovic** [comment]
Failure due to missing `shutdown` call. I need to think about if there is another way to destroy that d2sqlite handle without it complaining that it was done in a destructor..

**Comment by Geod24** [comment]
`ldc-latest-ci` is being weird:
```
source/agora/test/Ledger.d(59): [unittest] Node 0 has block height 40. Expected: 41
core.exception.AssertError@source/agora/test/Ledger.d(59): Node 0 has block height 40. Expected: 41
----------------
??:? _d_assert_msg [0x10d7c3566]
Ledger.d:59 @safe void agora.test.Ledger.__unittest_L29_C1().__lambda2!(ulong, geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI).RemoteAPI).__lambda2(ulong, geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI).RemoteAPI) [0x10d1497b9]
iteration.d:970 @safe std.typecons.Flag!("each").Flag std.algorithm.iteration.each!(agora.test.Ledger.__unittest_L29_C1().__lambda2).each!(std.range.enumerate!(ulong, geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI).RemoteAPI[]).enumerate(geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI).RemoteAPI[], ulong).Result).each(std.range.enumerate!(ulong, geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI).RemoteAPI[]).enumerate(geod24.LocalRest.RemoteAPI!(agora.test.Base.TestAPI).RemoteAPI[], ulong).Result) [0x10d14559c]
Ledger.d:58 void agora.test.Ledger.__unittest_L29_C1() [0x10d144d41]
Ledger.d:0 agora.test.Ledger.__unittest [0x10d149bf5]
??:? int core.runtime.runModuleUnitTests().__foreachbody2(object.ModuleInfo*) [0x10d7c82d7]
??:? int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))).__foreachbody2(ref rt.sections_elf_shared.DSO) [0x10d7f0991]
??:? int rt.sections_elf_shared.DSO.opApply(scope int delegate(ref rt.sections_elf_shared.DSO)) [0x10d7f1489]
??:? int rt.minfo.moduleinfos_apply(scope int delegate(immutable(object.ModuleInfo*))) [0x10d7f093b]
??:? int object.ModuleInfo.opApply(scope int delegate(object.ModuleInfo*)) [0x10d7de20e]
??:? runModuleUnitTests [0x10d7c807f]
??:? void rt.dmain2._d_run_main2(char[][], ulong, extern (C) int function(char[][])*).runAll() [0x10d7e863e]
??:? _d_run_main2 [0x10d7e847c]
??:? _d_run_main [0x10d7e82ec]
__entrypoint.d:8 main [0x10d159b84]
[1044BA11(----) INF] Doing network discovery..
[1044BA11(----) INF] Discovering from ["GA65EJ3JJQTNCIGO4LVVR3O5ZPEFX5ZP6Y6UMP5OO45JLPEKZIAPXAXF"]
[1044BA11(----) INF] Connecting to next set of addresses: ["GA65EJ3JJQTNCIGO4LVVR3O5ZPEFX5ZP6Y6UMP5OO45JLPEKZIAPXAXF"]
[1044BA11(----) INF] Establishing connection with GA65EJ3JJQTNCIGO4LVVR3O5ZPEFX5ZP6Y6UMP5OO45JLPEKZIAPXAXF...
[1044BA11(----) INF] Established new connection with peer: GA65EJ3JJQTNCIGO4LVVR3O5ZPEFX5ZP6Y6UMP5OO45JLPEKZIAPXAXF
[1044BA11(----) INF] Discovery reached. 1 peers connected.
[1044BCF1(----) INF] Doing network discovery..
[1044BCF1(----) INF] Discovering from ["GBZ6XPAHFWBBTX4KJG67TH2RQ7R6KZ2XAK476R2MA5NPBEHMZUKEXGZS"]
[1044BCF1(----) INF] Connecting to next set of addresses: ["GBZ6XPAHFWBBTX4KJG67TH2RQ7R6KZ2XAK476R2MA5NPBEHMZUKEXGZS"]
[1044BCF1(----) INF] Establishing connection with GBZ6XPAHFWBBTX4KJG67TH2RQ7R6KZ2XAK476R2MA5NPBEHMZUKEXGZS...
[1044BCF1(----) INF] Established new connection with peer: GBZ6XPAHFWBBTX4KJG67TH2RQ7R6KZ2XAK476R2MA5NPBEHMZUKEXGZS
[1044BCF1(----) INF] Discovery reached. 1 peers connected.
"Request timed-out"
"Request timed-out"
"Request timed-out"
"Request timed-out"
"Request timed-out"
"Request timed-out"
"Request timed-out"
"Request timed-out"
"Request timed-out"
"Request timed-out"
Couldn't retrieve blocks: Cannot reduce an empty input range w/o an explicit seed value.. Will try again later..
```

**Review by Geod24 — APPROVED** [review]

**Pull request merged**